kerastuner/engine/base_tuner.py

```diff
@@ -0,0 +1,232 @@
+                break
+            if trial.status == trial_module.TrialStatus.IDLE:
+                # Oracle is calculating, resend request.
+                continue
+
+            self.on_trial_begin(trial)
+            self.run_trial(trial, *fit_args, **fit_kwargs)
+            self.on_trial_end(trial)
+        self.on_search_end()
+
+    def run_trial(self, trial, *fit_args, **fit_kwargs):
+        raise NotImplementedError
+
+    def on_search_begin(self):
+        if self.logger:
+            self.logger.register_tuner(self.get_state())
+
+    def on_trial_begin(self, trial):
+        if self.logger:
+            self.logger.register_trial(trial.trial_id, trial.get_state())
+
+    def on_trial_end(self, trial):
+        self.oracle.end_trial(
+            trial.trial_id, trial_module.TrialStatus.COMPLETED)
+        self._display.on_trial_end(trial)
+        self._save_trial(trial)
+        self.save()
+
+    def on_search_end(self):
+        if self.logger:
+            self.logger.exit()
+
+    def get_best_models(self, num_models=1):
+        """Returns the best model(s), as determined by the tuner's objective.
+
+        Args:
+            num_models (int, optional): Number of best models to return.
+                Models will be returned in sorted order. Defaults to 1.
+
+        Returns:
+            List of trained model instances.
+        """
+        raise NotImplementedError
+
+    def search_space_summary(self, extended=False):
+        """Print search space summary.
+
+        Args:
+            extended: Bool, optional. Display extended summary.
+                Defaults to False.
+        """
+        display.section('Search space summary')
+        hp = self.oracle.get_space()
+        display.display_setting(
+            'Default search space size: %d' % len(hp.space))
+        for p in hp.space:
+            config = p.get_config()
+            name = config.pop('name')
+            display.subsection('%s (%s)' % (name, p.__class__.__name__))
+            display.display_settings(config)
+
+    def results_summary(self, num_trials=10):
+        """Display tuning results summary.
+
+        Args:
+            num_trials (int, optional): Number of trials to display.
+                Defaults to 10.
+            sort_metric (str, optional): Sorting metric, when not specified
+                sort models by objective value. Defaults to None.
+        """
+        display.section('Results summary')
+        display.display_setting('Results in %s' % self._get_project_dir())
+        best_trials = self.oracle.get_best_trials(num_trials)
+        display.display_setting('Showing %d best trials' % num_trials)
+        for trial in best_trials:
+            display.display_setting(
+                'Objective: {} Score: {}'.format(
+                    self.oracle.objective, trial.score))
+
+    @property
+    def remaining_trials(self):
+        return self.oracle.remaining_trials()
+
+    def get_state(self):
```

Comment:
Shouldn't `get_state` and `set_state` be implemented on the base class?

Reply:
I didn't think it should raise a `NotImplementedError`, but there's nothing stateful to be saved here. In `Tuner` we just save the `TunerStats` that says things like how many failed model builds there has been